accepted/0000-retail-pxe-database.md

```diff
@@ -0,0 +1,259 @@
+
+## Storing data in database
+
+As data are provided within salt event, implementing new event handler `SaltbootPxeUpdateEvent` with accompanying `SaltbootPxeUpdateMessageAction` should be sufficient.
+
+Event handler should also schedule generation of new PXE entry.
+
+## Exposing PXE entries to users
+
+In current implementation, consumer of the `pxe_update` event data is `pxe/terminal_entry` state which is part of PXE formula.
+
+In case of keeping the final action in salt space, there is a need to provide pillar data to the state:
+
+  * by supplying directly to state when scheduling pxe generating action.
+  * by exposing database data to salt pillar system.
+
+Exposing PXE data is not hard requirement but may have some benefit like enabling user to manually apply `pxe/terminal_entry` or similar state to (re)generate pxe entries. This option can save resources by making UI/XML-RPC API changes non-essential.
+
+Database can be exposed to pillar system by external pillar mechanism connecting to the database using postgresql external pillar:
+
+```
+postgres:
+  db: susemanager
+  host: localhost
+  pass: spacewalk
+  port: 5432
+  user: spacewalk
+
+ext_pillar:
+  - postgres:
+      'retail_terminals':
+         query: "SELECT S.name, root_device, salt_device, boot_image, kernel_parameters,mac_address FROM
+                (SELECT * FROM suseRetailMinionInfo as RMI, rhnServer as S WHERE RMI.pxeserver_id = S.id AND S.name LIKE %s) AS I JOIN rhnServer AS S ON I.server_id = S.id"
+         as_list: False
+         depth: 1
+```
+
+or by writing pillar files for given PXE server using already existing Java methods provided by `MinionPillarFileManager`.
+
+
+Keeping data hidden in database has a positive side by not cluttering pillar space with rarely used data.
+To allow user to manually trigger PXE entries regeneration it is needed to develop UI or XMLRPC API endpoint, proposed below.
+
+Using these data `pxe/terminal_entry` state would need to be adapted to look for pillar data in correct places and also to support generating single pxe entry or all pxe entries depending on the values passed to the state.
+
+## Generating PXE entries
+
+Update to routine is needed to support terminals connected directly to Uyuni server. Mainly calling `pxe/terminal_entry` state using salt runner or other means locally on Uyuni.
+
+Complete routine should look like:
+
+![Generating PXE entries flowchart](images/0000-retail-pxe-db-workflow.png)
+
+## Updating PXE entries
+
+Moving terminal between pxe servers happen. This means event handler needs to be aware of this possibility and adapt `pxeserver_id` field as needed.
+  
+Removal of the PXE server (branch server) system profile would set `pxeserver_id` to `null`. To update the entry terminal will need to be rebooted under new PXE server or user would need to manually update using API call.
```

Comment:
I feel we will have to give a `NULL` `pxeserver_id` a precise meaning - I can see we run into the situation where it's not immediately clear from logs whether a terminal is served by an unknown PXE server from one which is served by the Server itself.

What benefits do you see in using a system ID rather than, say, the fully-qualified domain name as a string?

Reply:
If we use FQDN passed from terminal, we will sooner or later need to determine if it is known system, uyuni server or something else anyway. Because we want to know if we should schedule salt state, salt-call or nothing or something else. That boils to question what we want to do in case of not managed pxe server?

I see it that upon receiving the event we do server lookup based on FQDN received in the event. If we find it - assign `pxeserver_id`, if not we will check if it is uyuni server FQDN and set `NULL`. If it is not even uyuni server FQDN, report warning in the log and set `NULL`

Then when `pxeserver_id` is `NULL` we will always generate pxe entries on Uyuni server. This may be documented and users can use this even in case when using non managed pxe server and copy the files over there.